### PR TITLE
KAFKA-14683 Migrate WorkerSinkTaskTest to Mockito (2/3)

### PR DESCRIPTION
**Pull request by hgeraldino**
Second batch of WorkerSinkTaskTest unit tests migrated to Mockito

### CODE REVIEW

**Comment by hgeraldino** [comment]
Thanks @gharris1727 and @C0urante for reviewing and merging the [first PR](https://github.com/apache/kafka/pull/14663)

The second second (of three) batches is ready for review now after rebasing :)

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSinkTaskMockitoTest.java [hunk 1]
```diff
@@ -343,6 +354,103 @@ public void testShutdown() throws Exception {
...
+        workerTask.initialize(TASK_CONFIG);
+        workerTask.initializeAndStart();
+        verifyInitializeTask();
+
+        expectPollInitialAssignment()
+                // If a retriable exception is thrown, we should redeliver the same batch, pausing the consumer in the meantime
+                .thenAnswer(expectConsumerPoll(1))
+                // Retry delivery should succeed
+                .thenAnswer(expectConsumerPoll(0))
+                .thenAnswer(expectConsumerPoll(1))
```
Comment (gharris1727) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSinkTaskMockitoTest.java]: The original test didn't have this additional record, and the current test passes without it.
The test also has 4 iteration() calls, which should be:
1. initial assignment
2. first record
3. after pause, redelivery
4. after request commit

I think there should only be 3 thenAnswer calls here, and the expectConsumerPoll(1) is the one that should be removed.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSinkTaskMockitoTest.java [hunk 2]
```diff
@@ -343,6 +354,103 @@ public void testShutdown() throws Exception {
         verify(headerConverter).close();
     }
 
+    @Test
+    public void testPollRedelivery() {
+        createTask(initialState);
+        expectTaskGetTopic();
+
+        when(consumer.assignment()).thenReturn(INITIAL_ASSIGNMENT);
+        INITIAL_ASSIGNMENT.forEach(tp -> when(consumer.position(tp)).thenReturn(FIRST_OFFSET));
```
Comment (gharris1727) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSinkTaskMockitoTest.java]: This is already in expectPollInitialAssignment, right?
```suggestion
```

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSinkTaskMockitoTest.java [hunk 3]
```diff
@@ -482,21 +590,187 @@ public void testPartialRevocationAndAssignment() {
...
+        offsets.put(TOPIC_PARTITION, new OffsetAndMetadata(FIRST_OFFSET + 2));
+        offsets.put(TOPIC_PARTITION2, new OffsetAndMetadata(FIRST_OFFSET));
+        when(sinkTask.preCommit(offsets)).thenReturn(offsets);
+
+        workerTask.execute();
+
+        // stop wakes up the consumer
+        verify(consumer).wakeup();
+
+        verify(sinkTask).close(any(Collection.class));
```
Comment (gharris1727) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSinkTaskMockitoTest.java]: You can remove this unchecked supression
```suggestion
        verify(sinkTask).close(any());
```

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSinkTaskMockitoTest.java [hunk 4]
```diff
@@ -482,21 +590,187 @@ public void testPartialRevocationAndAssignment() {
...
+        TaskConfig taskConfig = new TaskConfig(props);
+
+        createTask(TargetState.PAUSED);
+
+        workerTask.initialize(taskConfig);
+        workerTask.initializeAndStart();
+
+        ArgumentCaptor<Pattern> topicsRegex = ArgumentCaptor.forClass(Pattern.class);
+
+        verify(consumer).subscribe(topicsRegex.capture(), rebalanceListener.capture());
```
Comment (gharris1727) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSinkTaskMockitoTest.java:758]: Could you add an assertion on the topicsRegex?
```suggestion
        verify(consumer).subscribe(topicsRegex.capture(), rebalanceListener.capture());
        assertEquals("te.*", topicsRegex.getValue().pattern());
```

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSinkTaskMockitoTest.java [hunk 5]
```diff
@@ -558,6 +832,143 @@ public void testMetricsGroup() {
...
+                            new ConsumerRecord<>(TOPIC, PARTITION, FIRST_OFFSET + recordsReturnedTp1 + 2, RecordBatch.NO_TIMESTAMP, TimestampType.NO_TIMESTAMP_TYPE,
+                                    0, 0, keyB.getBytes(), valueB.getBytes(encodingB), headersB, Optional.empty())
+                    );
+                    return new ConsumerRecords<>(Collections.singletonMap(new TopicPartition(TOPIC, PARTITION), records));
+                });
+
+        expectTransformation(null);
+
+        workerTask.iteration(); // iter 1 -- initial assignment
+        workerTask.iteration(); // iter 2 -- deliver 1 record
```
Comment (gharris1727) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSinkTaskMockitoTest.java]: the original comment is wrong
```suggestion
        workerTask.iteration(); // iter 2 -- deliver records
```

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSinkTaskTest.java [hunk 6]
```diff
@@ -162,7 +153,6 @@ public class WorkerSinkTaskTest {
     @Mock
     private ErrorHandlingMetrics errorHandlingMetrics;
```
Comment (gharris1727) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSinkTaskTest.java:153]: Above here, taskId1 is not used

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSinkTaskMockitoTest.java [hunk 7]
```diff
@@ -558,6 +832,143 @@ public void testMetricsGroup() {
...
+        // Setting up Worker Sink Task to check metrics
+        workerTask = new WorkerSinkTask(
+                taskId, sinkTask, statusListener, TargetState.PAUSED, workerConfig, ClusterConfigState.EMPTY, metrics,
+                keyConverter, valueConverter, errorHandlingMetrics, headerConverter,
+                transformationChain, mockConsumer, pluginLoader, time,
+                RetryWithToleranceOperatorTest.noopOperator(), null, statusBackingStore, Collections::emptyList);
+        mockConsumer.updateBeginningOffsets(
+                new HashMap<TopicPartition, Long>() {{
+                    put(TOPIC_PARTITION, 0 * 1L);
+                    put(TOPIC_PARTITION2, 0 * 1L);
```
Comment (gharris1727) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSinkTaskMockitoTest.java]: This is a warning in my IDE and the multiply doesn't seem to serve any purpose.
```suggestion
                    put(TOPIC_PARTITION, 0L);
                    put(TOPIC_PARTITION2, 0L);
```

**Review by gharris1727 — COMMENTED** [review]
Thanks @hgeraldino! I only had minor comments, it looks great!

Also thanks for addressing the Whitebox usage, I think the immutable lastCommittedOffsets is probably better than making the field itself package-private.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSinkTaskTest.java (hunk 6 above)
Comment (hgeraldino) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSinkTaskTest.java:153]: removed

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSinkTaskMockitoTest.java (hunk 1 above)
Comment (hgeraldino) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSinkTaskMockitoTest.java]: yep, that makes sense.

removed

**Review by gharris1727 — APPROVED** [review]
LGTM, thanks @hgeraldino!

**Comment by gharris1727** [comment]
Test failures appear unrelated, and the runtime tests pass locally for me.

**Pull request merged**